### ISSUE TITLE
Analytically Continuing SrVO3 using SOM

### ISSUE THREAD

**Issue by AlynJ** [krivenko/som#4]
Dear Dr. Krivenko,

I’m interested in using your SOM library to perform analytic continuation calculations on the Green’s functions and self energies generated in the dft tools module of TRIQS. I’m attempting to follow the DFT+DMFT calculation of SrVO3 using the parameters and guide given on the dft tools website. I would like to calculate the real frequency self-energy to be used in the post processing part of the dft tools module. 

However, when I first try to use SOM to analytically continue the lattice Green’s function to determine the DoS, I’m unable to get the Hubbard bands. This can be seen in the figure of the spectral function provided below. I’ve also included the spectral function determined from the Pade approximation given in the TRIQS library and the analytic continuation calculated from the ΏMaxEnt program [1]. Would you please help me to see where I’m going wrong in using the SOM library?

 
![image](https://user-images.githubusercontent.com/34216043/33880243-ee18bb24-df28-11e7-8b4b-c42fc9c32e49.png)

When I ran ‘make test’ after compiling SOM, the solution_worker test returned ***EXCEPTION (SEGFAULT). However, when I did the example on the ‘Fermionic Green’s function or self-energy’ section of the SOM site, I reproduced the same analytically continued spectral function. Could this be causing the aforementioned discrepancy?

Kind regards,

Alyn

P.S. - I’ve provided my SrVO3.h5, SOM python script, CMakeCache.txt and LastTest.log files below. If there’s any further information you require, please let me know!

References
[1] – D. Bergeron and A-M. S. Tremblay, Physical Review E 94.2 (2016): 023303. (
 https://arxiv.org/pdf/1507.01012.pdf )

[Som_query_files_AlynJ.zip](https://github.com/krivenko/som/files/1551209/Som_query_files_AlynJ.zip)

**Comment by MHarland** [comment]
Hi Alyn,

assuming you had set the default parameters it looks like your run lacked ergodicity. Larger values for the parameters f and t could fix this. I would start with trying different values for f, e.g. up to 2000. The continuation will take significantly more computation time, though.

**Comment by AlynJ** [comment]
Thank you for your quick response!

After increasing f to 2000, this did indeed improve the results. The upper hubbard band appears, but at a higher energy than expected (I've attached the graph below). I’m going to try to increase it again to see if I get better results and I’ll let you know. As this increased the run time, are there other parameters which I can reduce to decrease the run time and still get useful results? 

I’ve got a few question about SOM to help my understanding of the software. How does the software determine the real part of the analytically continued Green’s function? Does it use the Kramers-Kronig relation? 

To analytically continue the self energy calculated in dft tools, should I use the tail parameters (Sigma.tail) for the moments? Would it be better to create an auxiliary Green’s function from the self-energy and then analytically continue that instead?

Many thanks,

Alyn

![an_con_comp_fig_f 2000_t 100_l 512](https://user-images.githubusercontent.com/34216043/33995244-45742a54-e0d5-11e7-8721-f3cc3439dcb4.png)

**Comment by krivenko** [comment]
Dear Alyn,

Thank you for your interest in our software.

> When I ran ‘make test’ after compiling SOM, the solution_worker test returned ***EXCEPTION (SEGFAULT).

Okay, I am going to have a closer look at this issue when I have more time.
I cannot reproduce the failure on my machine, even though I recently saw it with a different build configuration. What is your compiler? `CMakeCache.txt` says it is `mpicxx`, which is normally the MPI wrapper around `g++`, `clang++`, `icpc`, etc.

Anyhow, I highly doubt that this test failure affects your actual results in a dramatic way.

> Would you please help me to see where I’m going wrong in using the SOM library?

I agree with Malte, there seems to be an ergodicity problem that might be overcome by increasing parameters `f` and `t`.

Also, you should always look at the reconstructed GF curve (the 'Fermionic Green’s function or self-energy' example shows how to do that). Significant deviations of the reconstructed GF from the input data signal a serious problem in the continuation procedure, and render obtained results useless.

> As this increased the run time, are there other parameters which I can reduce to decrease the run time and still get useful results?

You could reduce 'l' to get a qualitatively similar but more noisy curve.

> How does the software determine the real part of the analytically continued Green’s function? Does it use the Kramers-Kronig relation?

It constructs real and imaginary parts of the retarded Green's function from the computed spectral function using the last equation in http://krivenko.github.io/som/kernels.html#fermionic-thermal-green-s-function.

> To analytically continue the self energy calculated in dft tools, should I use the tail parameters (Sigma.tail) for the moments?

dft_tools is something I have never used. But yes, you must somehow find the coefficient in front of `1/iw_n` in the high-frequency expansion of the self-energy.

> Would it be better to create an auxiliary Green’s function from the self-energy and then analytically continue that instead?

I have no personal experience with this sort of approaches, so you may want to try both ways.

Igor

**Comment by AlynJ** [comment]
Hi all!

Thanks for the response! The compilers I’m using are GCC-7.1.0 with openmpi-3.0.0 wrappers.

I’ve increased f to 10000, t to 150 and reduced l to 500 (to reduce calculation time). I get the 3 peak structure (see below), but the peaks appear at different energies than ΩMaxEnt. Any suggestions to why that is? I’ve also included the reconstructed GF curve.

Many thanks,

Alyn

![an_con_comp_fig_f 10000_t 150_l 500](https://user-images.githubusercontent.com/34216043/34104686-6c8f3bb2-e3e9-11e7-80da-5f8532fd105a.png)

![an_con_comp_fig_rec_f 10000_t 150_l 500](https://user-images.githubusercontent.com/34216043/34104691-72347794-e3e9-11e7-87a2-658f1a6f80c8.png)

**Comment by krivenko** [comment]
> I get the 3 peak structure (see below), but the peaks appear at different energies than ΩMaxEnt. Any suggestions to why that is?

Well, analytical continuation procedures are rather unreliable when it comes to high-frequency features. It may very well be that both solutions fit the input data within the same precision.

Try to reconstruct GF from the ΩMaxEnt spectral function, maybe it will give you an insight into quality of the solutions.

**Comment by krivenko** [comment]
@AlynJ Please, check that solution_worker test is fixed by recent commits.

**Comment by AlynJ** [comment]
Hi all,

I've rebuilt the SOM library with the updates and now the solution_worker test passes!

I've also reconstructed G(iω) from the ΩMaxEnt spectral function (shown below) and there is a discrepancy in the real part of G(iω) at low imaginary frequencies. Would it be correct to think that this discrepancy is the cause of the Hubbard band peaks being located at different energies? 

![an_con_comp_fig_rec_omegamaxent](https://user-images.githubusercontent.com/34216043/34178919-ce88b4ca-e500-11e7-8e0e-1a415c93e43f.png)

As you have probably guessed, I'm new to using these different types of analytical continuation software!

Many thanks,

Alyn

**Comment by krivenko** [comment]
Hi,

> I've also reconstructed G(iω) from the ΩMaxEnt spectral function (shown below) and there is a discrepancy in the real part of G(iω) at low imaginary frequencies. Would it be correct to think that this discrepancy is the cause of the Hubbard band peaks being located at different energies?

The discrepancy you are speaking about means that ΩMaxEnt results are not to be trusted close to \omega=0. Nonetheless, it is still hard to draw any conclusions regarding the Hubbard bands.

**Comment by AlynJ** [comment]
Hi,

Apart from reconstructing G(iω), are there other ways to test whether I can be confident in my results I am getting from the different analytical continuation software? 

I don't understand how I would determine which of my analytically continued spectral function is more reliable. For example, my ΩMaxEnt spectral function better matches previous calculations of DFT+DMFT SrVO3 spectral functions, but the reconstructed G(iω) is worse at matching the actual G(iω).

Thank you for answering my questions,

Alyn

**Comment by krivenko** [comment]
> Apart from reconstructing G(iω), are there other ways to test whether I can be confident in my results I am getting from the different analytical continuation software?

I have to admit there are none. Numerical analytical continuation is not something one can be confident in. Try to find an estimate of the Hubbard band position that is not based on continuation (Exact Diagonalization results or something similar).

> I don't understand how I would determine which of my analytically continued spectral function is more reliable.

I doubt that anyone could give you a definite answer to this question. In my personal opinion, the discrepancy in the reconstructed G(iω) is way too big for the MaxEnt results being useful. However, this does not necessarily mean SOM's solution is correct.

> For example, my ΩMaxEnt spectral function better matches previous calculations of DFT+DMFT SrVO3 spectral functions

Those previous spectral functions are very likely to have been obtained with some sort of analytical continuation method, which means their validity is also disputable.

**Comment by AlynJ** [comment]
Hi,

Thank you very much for taking the time to answer all of these questions about SOM and analytical continuation in general, this has clarified many areas which I was unsure of. You have answered all the questions I currently have about this. I believe I've got a better appreciation of analytical continuation now. 

I should mention for the benefit of any other users looking at this thread that my results don't represent the full scope of the software I used. There may be ways that these software may better optimise the results I calculated.

Many thanks,

Alyn
